### ISSUE TITLE
Datastore Entity.Builder.Set don't accept nulls

### ISSUE THREAD

**Issue by MTocchettoP** [GoogleCloudPlatform/google-cloud-java#3583]
Hi,

I've ran into this problem a while a go and although it is not a bug it does make the API harder to use.

The problem: `set` don't take null values
The solution: Use `setNull` or `set` and set a Value.Null (something like that) 
The solution's problem: This require the api client to manually check if a value is null in order to call the proper `set` method.

Here is a sample code
```
public void put(BatchExecution obj) {
    Key key = keyFactory.newKey(obj.getId());           
    FullEntity<Key> incBEEntity = Entity.newBuilder(key)  
        .set(BatchExecution.ID, obj.getId())
        .set(BatchExecution.NAME, obj.getName())
        .set(BatchExecution.CREATETIME, obj.getCreateTime())
        .set(BatchExecution.ELAPSEDTIME, obj.getElapsedTime()) //I break the code because elapesedTime was never set in the object
        .set(BatchExecution.STATUS, obj.getStatus().name())
        .build();
    datastore.put(incBEEntity);
  }
```

This looks very weird to me since when you call  `set` on a null value eventually the stack call throws an exception on a method called `CheckNull`. So, since a check is eventually made down the stack, why not move this null check to the overloaded `set` methods and if it yields true the method will redirect the request to the `setNull` method. Having two different `set` looks like unnecessary exposure of API innards work. If the API handles lets say a String differently than a Null, it's the API's problem to deal with that, not the client. This concern can be seen in the overload of the `set` method. The API need to treat a String differently from a TimeStamp (obviously), but this is not exposed to the client, it`s all the same `set` method.

In sum, my request is: Is it possible to have `set` accepting null values instead of having the client  manually check if an object is null and calling  the appropriate method?

**Comment by ChrisSorto89** [comment]
The way the Entity.Builder Set and SetNull methods function is intentional.  This is a widely used API and it is expected that the client decides how they want to handle Nulls within there own application.  This allows the client to treat Nulls as either valid or invalid input, depending on there particular process and system.  This avoids any clients from accidentally storing Nulls if that is not what they expect.  If you would like to create a set method that is reusable and accepts nulls, it would be possible to do this through inheritance(extends) and overriding under a new class and API.  Please note, that you must create a new set method for each different datatype you wish to add this logic to.  I have given an example of how this can be done below:

class Entity.BuilderOverride extends Entity.Builder
{ 
    // This method overrides set for strings.  It must be done for each different data type wish to handle
    public B set(String name, String value)
             {
               if(value==null) 
               {
                 setNull(name);
                }
            else
            {
              super.set(name,value);
              }
         }
}

**Comment by MTocchettoP** [comment]
Yes, I ended up going with a somewhat similar solution. I've refactored my datastore client class (which is basically my dao layer) to work generically, despite what's being stored. I used generics and custom annotations that provide a default value if the attribute is null. I get the idea of giving the client the power to decide what to do with Null values, but I will argue that this is a problem for the client's business layer and not the DAO. Sure, assuming I don't want ever to persist nulls I could do the checks on my business layer and ensure my DAO only get complete object and as such I could use the code I mentioned on the first post, but then again you have a vulnerable DAO that will break if some other class that's not aware of this requirement don't treat null accordingly. Alas, I still find the implementation weird, because even if I do want to set nulls, I have to check it and call a different method in the DAO, which should not be his job, the data layer does not/ should not, care for the data, all it should care about is how to interface with the database, what is being passed to it is not it's problem.

**Comment by satishgadepally** [comment]
We can create new method inside builder class called setVal.

Now instead of calling set method we can call setVal method.

Internally setVal method will check for null and according to it it will call set or setNull method.

Here we can describe setVal method as below.
 

setVal(String param1, Object param2)
{
If(param2  == null)
{
SetNull(param1);
}
else{
set(param1,param2);
}
}

**Comment by NAGAMANIKONDA** [comment]
In com.google.cloud.datastore.Value 
If no other implications are there we can remove the **checkNotNull**

But it will be good to externalize this like a property to leave it to clients for storing null values or not

I understand that there are some specific data types like lat long, time stamps etc which will not make sense to be null 


    @Override
    public B set(V value) {
//      this.value = checkNotNull(value);
      return self();
    }
